jerry-core/vm/vm.c

```diff
@@ -2613,16 +2668,33 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
         }
         case VM_OC_EXT_RETURN:
         {
-          result = left_value;
-          left_value = ECMA_VALUE_UNDEFINED;
-
           ecma_value_t *stack_bottom_p = VM_GET_REGISTERS (frame_ctx_p) + register_end + frame_ctx_p->context_depth;
 
           while (stack_top_p > stack_bottom_p)
           {
-            ecma_fast_free_value (*(--stack_top_p));
+            ecma_value_t stack_item = *(--stack_top_p);
```

Comment:
I would do this differently. I would push a normal context to the stack, and store the previous value of `context_depth` as part of this context. Then set `context_depth` to the top of the stack. When `vm_stack_context_abort` runs, it does not only remove the context, but also free the JS values stored before this context is pushed.